libshaderc_spvc/src/spvcir_pass.cc

```diff
@@ -14,25 +14,189 @@
+      const std::string extension_name =
+          reinterpret_cast<const char *>(inst->GetInOperand(0u).words.data());
+      ir_->declared_extensions.push_back(std::move(extension_name));
+      break;
+    }
+
+    case SpvOpMemoryModel: {
+      ir_->addressing_model =
+          static_cast<spv::AddressingModel>(inst->GetSingleWordInOperand(0u));
+      ir_->memory_model =
+          static_cast<spv::MemoryModel>(inst->GetSingleWordInOperand(1u));
+      break;
+    }
+
+    case SpvOpEntryPoint: {
+      auto function_id = inst->GetSingleWordInOperand(1u);
+      auto execution_mode =
+          static_cast<spv::ExecutionModel>(inst->GetSingleWordInOperand(0u));
+
+      const char *entry_name =
+          reinterpret_cast<const char *>(inst->GetInOperand(2u).words.data());
+
+      auto itr = ir_->entry_points.insert(std::make_pair(
+          function_id, spirv_cross::SPIREntryPoint(function_id, execution_mode,
+                                                   entry_name)));
+      auto &e = itr.first->second;
+      ir_->set_name(function_id, e.name);
+
+      for (uint32_t i = 3; i < inst->NumInOperands(); ++i) {
+        e.interface_variables.push_back(inst->GetSingleWordInOperand(i));
+      }
+
+      if (!ir_->default_entry_point) ir_->default_entry_point = function_id;
+      break;
+    }
+
+    case SpvOpTypeVoid: {
+      auto id = inst->result_id();
+      auto &type = set<spirv_cross::SPIRType>(id);
+      type.basetype = spirv_cross::SPIRType::Void;
+      break;
+    }
+
+    case SpvOpTypeFunction: {
+      auto id = inst->result_id();
+      auto result = inst->type_id();
+
+      // reading Parameter 0 Type, Parameter 1 Type, ...
+      auto &func = set<spirv_cross::SPIRFunctionPrototype>(id, result);
+      for (uint32_t i = 2; i < inst->NumInOperands(); ++i) {
+        func.parameter_types.push_back(inst->GetSingleWordInOperand(i));
+      }
+
+      break;
+    }
+
+    case SpvOpFunction: {
+      auto result = inst->type_id();
+      auto id = inst->result_id();
+      auto type = inst->GetSingleWordInOperand(1u);
+
+      assert(!current_function_ &&
+             "Must end a function before starting a new one!");
+
+      current_function_ = &set<spirv_cross::SPIRFunction>(id, result, type);
+      break;
+    }
+
+      // Blocks
+    case SpvOpLabel: {
+      // OpLabel always starts a block.
+      assert(current_function_ && "Blocks cannot exist outside functions!");
+
+      uint32_t id = inst->result_id();
+
+      current_function_->blocks.push_back(id);
+
+      if (!current_function_->entry_block) {
+        current_function_->entry_block = id;
+      }
+      assert(!current_block_ &&
+             "Cannot start a block before ending the current block.");
+
+      current_block_ = &set<spirv_cross::SPIRBlock>(id);
+      break;
+    }
+
+    case SpvOpReturn: {
+      assert(current_block_ && "Trying to end a non-existing block.");
+      current_block_->terminator = spirv_cross::SPIRBlock::Return;
```

Comment:
Eventually I can see this being refactored into TerminateBlock( ... terminator type ...), which also resets current_block_ to nullptr.

But not yet, since you should only do that after having at least one more terminator case.